src/main/java/ua/com/juja/microservices/keepers/slackbot/controller/KeepersSlackCommandController.java

```diff
@@ -18,77 +26,136 @@
-                                           @RequestParam("text") String text) {
-        logger.debug("Received slash command GetKeeperDirections: from user: [{}] command: [{}] token: [{}]",
-                fromUser, text, token);
-
-        if (!token.equals(slackToken)) {
-            logger.warn("Received invalid slack token: [{}] in command Keeper for user: [{}]", token, fromUser);
-            return new RichMessage("Sorry! You're not lucky enough to use our slack command.");
+    public void getKeeperDirections(@RequestParam("token") String token,
+                                   @RequestParam("user_name") String fromUser,
+                                   @RequestParam("text") String text,
+                                   @RequestParam("response_url") String responseUrl,
+                                   HttpServletResponse httpServletResponse) {
+        try {
+            LOGGER.debug("Received slash command GetKeeperDirections: from user: [{}] command: [{}] token: [{}]",
+                    fromUser, text, token);
+
+            if (!token.equals(slackToken)) {
+                LOGGER.warn("Received invalid slack token: [{}] in command Keeper for user: [{}]", token, fromUser);
+                sendQuickResponse(httpServletResponse, SORRY_MESSAGE);
+            } else {
+                sendQuickResponse(httpServletResponse, IN_PROGRESS);
+                String response = keeperService.getKeeperDirections(fromUser, text);
+                LOGGER.info("GetKeeperDirections command processed : user: [{}] text: [{}] and sent response to slack: [{}]",
+                        fromUser, text, response);
+                sendDelayedResponse(responseUrl, response);
+            }
+        } catch (BaseBotException bex){
+            sendBBEMessage(responseUrl, bex);
+        } catch (Exception ex){
+            sendEMessage(responseUrl, ex);
         }
+    }
 
-        String response = keeperService.getKeeperDirections(fromUser, text);
+    private void sendQuickResponse(HttpServletResponse httpServletResponse, String message) throws IOException {
+        httpServletResponse.setStatus(HttpServletResponse.SC_OK);
+        try (PrintWriter printWriter = httpServletResponse.getWriter()) {
+            printWriter.print(message);
+            printWriter.flush();
+        }
+        LOGGER.info("Sent a quick response with message '{}'", message);
+    }
 
-        logger.info("GetKeeperDirections command processed : user: [{}] text: [{}] and sent response to slack: [{}]",
-                fromUser, text, response);
+    private void sendDelayedResponse(String responseUrl, String response) {
+        String slackAnswer = restTemplate.postForObject(responseUrl, new CustomRichMessage(response), String.class);
+        LOGGER.info("Slack answered: [{}]", slackAnswer == null ? "null" : slackAnswer);
+    }
 
-        return new RichMessage(response);
+    private void sendBBEMessage(String responseUrl, BaseBotException bex) {
```

Comment:
please rename to full method name because BBE and E are not obvious abbreviations.

Reply:
ok

Reply:
fixed